src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Microsoft.VisualStudio.LanguageServerClient.Razor.csproj

```diff
@@ -17,6 +17,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="System.Composition" Version="$(SystemCompositionPackageVersion)" />
```

Comment:
Why's this needed?

Reply:
We were getting that indirectly via the Microsoft.CodeAnalysis dependency. Now that that's gone, I had to add a direct reference. Some stuff (e.g. ExportAttribute) doesn't seem to be a part of 4.7.2. See [https://docs.microsoft.com/en-us/dotnet/api/system.composition.sharedattribute?view=dotnet-plat-ext-3.1&viewFallbackFrom=netframework-4.7.2](https://docs.microsoft.com/en-us/dotnet/api/system.composition.sharedattribute?view=dotnet-plat-ext-3.1&viewFallbackFrom=netframework-4.7.2)

---
In reply to: [447116379](https://github.com/dotnet/aspnetcore-tooling/pull/2137#discussion_r447116379) [](ancestors = 447116379)

Reply:
(FWIW in case others are confused, System.Composition != System.ComponentModel.Composition.  It's NuGet MEF vs .NET MEF.  Both separately define `[ExportAttribute]` but the behavior varies, hence the need to preserve the previously-transitive reference.)

Reply:
Right, they are different things and defined same classes in different namespaces. Also, I'm not seeing [Shared] (not [Export] as I erroneously said before) in 4.7.2 at all - seems to require a nuget package regardless. 

---
In reply to: [447242227](https://github.com/dotnet/aspnetcore-tooling/pull/2137#discussion_r447242227) [](ancestors = 447242227)

Reply:
That's the difference.  In .NET MEF, the lifetime or sharing boundary is specified with an optional `[PartCreationPolicyAttribute(...)]` (there's an implicit default policy).  In NuGet MEF, you must explicitly specify the lifetime or sharing boundary, e.g. with `[SharedAttribute]` (which only exists in the System.Composition assembly), hence the need for the package. :)